File: Code/decoder.py
from itertools import cycle, product, tee, islice
import copy
import numpy
import operator
import time
import random
#######################################################
#         Decoder that takes from SWE file and 
#       populates list and variables accordingly 
#######################################################
start_time = time.time()
f = open("tests/test06.swe", "r")
k = int(f.readline())
s = f.readline().strip()
t_strings = []
for x in range(0, k):
    t_strings.append(f.readline().strip())
myline = f.readline().strip()
dict_of_R = {}
unused_Rs = []
unused_Rs_copy = []
while(myline):
    tmp_line = myline.split(":")
    unused_Rs.append(tmp_line[0])
    unused_Rs_copy.append(tmp_line[0])
    tmp_list_items = tmp_line[1].split(",")
    tmp_list = []
    for x in tmp_list_items:
        if x in s:
            tmp_list.append(x)
    dict_of_R[tmp_line[0]] = tmp_list
    myline = f.readline().strip()
f.close()

t_strings_copy = list(dict.fromkeys(t_strings))
t_strings_new = copy.deepcopy(t_strings_copy)

t_strings_copy.sort(key=len, reverse=False)
for n in t_strings_copy:
    for m in range(0, len(t_strings_copy)):
        if n in t_strings_copy[m]:
            if n != t_strings_copy[m]:
                if n in t_strings_new:
                    t_strings_new.remove(n)

# Lets start by cutting out all r elements in each R by checking if they alone are a substring of s
# thus eliminating all r's that could never form a substring of s
# and also get rid of all R that are unused (never appear in any of our t strings)
dict_of_R_copy = {}
for element in dict_of_R.keys():
    dict_of_R_copy[element]=dict_of_R[element][0]
used_R =[]
for x in t_strings_new:
    tmp_len = len(x)
    for i in range(0, tmp_len):
        if(x[i].isupper()):
            if x[i] in unused_Rs:
                unused_Rs.remove(x[i])
            if x[i] not in used_R:
                used_R.append(x[i])
for x in unused_Rs:
    del dict_of_R[x]


def expansion(curr_string, big_Rs, choice):
    new_string = ""
    for c in curr_string:
        if c.isupper():
            tmpindex = big_Rs.index(c)
            new_string += choice[tmpindex]
        else: 
            new_string += c       
    return new_string
    

dictionary_items = dict_of_R.items()
all_lists = []
for item in dictionary_items:
    if "" in item[1]:
        item[1].remove("")
    all_lists.append(numpy.array(item[1]))


#######################################################
#      Algorithm that is used to find if soultion
#    exists and returns it if it does otherwise NO
#######################################################
def listoflists(all_lists1, t_string, used_R, s):
    count_t_strings = len(t_string)
    for x in product(*all_lists1):
        counter = 0
        for y in t_string:
            current_t_string = expansion(y, used_R, x)
            if current_t_string not in s:
                break
            else:
                counter += 1 
        if counter == count_t_strings:
            for d in unused_Rs_copy:
                if d in used_R:
                    indx = used_R.index(d)
                    print("{}: {}".format(d, x[indx]))
                else:
                    print("{}: {}".format(d, dict_of_R_copy[d]))
            return
    print('NO')

t_strings_new.sort(key=len, reverse=True)
used_R.sort()

listoflists(all_lists, t_strings_new, used_R, s)

print("--- {:.2f} seconds ---".format(time.time() - start_time))

File: Code/decoder.cpp
#include <iostream>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <map>
#include <cctype>
#include <cstring>
#include <vector>
#include <algorithm>
#include<sstream>

using namespace std;

//string expansion(string curr_string, string big_Rs[], string choice[]){  
//}

int main()
{
    //char letters[26] = {'a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z'}
    //char capitalLetters[26] = {'A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z'}
    //string valid input = '';

    int k;
    cin >> k;

    string s;
    cin >> s;

    vector<string> t_strings;;
    for(int i = 0; i < k; i++){
        string tmpstr;
        cin >> tmpstr;
        t_strings.push_back(tmpstr);
    }

    vector<string> unused_Rs;
    map<string, vector<string>> dict_of_R;

    while(!cin.fail()){
        string myline;
        cin >> myline;
        vector<string> result;
        stringstream s_stream(myline); //create string stream from the string
        string R;
        getline(s_stream, R, ':');
        unused_Rs.push_back(R);
        while(s_stream.good()) {
            string substr;
            getline(s_stream, substr, ','); //get first string delimited by comma
            result.push_back(substr);
        }
        dict_of_R[R] = result;
    }

    vector<string> unused_Rs_copy = unused_Rs;

    map<string, vector<string>> dict_of_R_copy = dict_of_R;
    vector<string> used_R;

    for (size_t i = 0; i < t_strings.size(); i++){
        int tmp_len = t_strings[i].length();
        for(int j = 0; j < tmp_len; j++){
            if(isupper(t_strings[i][j])){
                if(find(unused_Rs.begin(), unused_Rs.end(), t_strings[i][j]) != unused_Rs.end()){
                    remove(unused_Rs.begin(), unused_Rs.end(), t_strings[i][j]);
                }
                if(!find(used_R.begin(), used_R.end(), t_strings[i][j]) != used_R.end()){
                    used_R.push_back(t_strings[i][j])
                }
                vector<string> tmp_dict_list = dict_of_R[t_strings[i][j]];
                int tmp_dict_list_len = dict_of_R[t_strings[i][j]].size();
                for(int k = 0; k < tmp_dict_list_len; k++){
                    if (s.find(tmp_dict_list[k]) == string::npos) {
                        remove(dict_of_R[t_strings[i][j]].begin(), dict_of_R[t_strings[i][j]].end(), tmp_dict_list[k]);
                    } 
                }
            }
        }
	}

    for(size_t l = 0; l < unused_Rs.size(); l++){
        dict_of_R.erase(unused_Rs[l]);
    }

    cout << dict_of_R << endl;
    
    return 0;
}